activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb

```diff
@@ -914,7 +913,7 @@ def add_pg_decoders
             "timestamptz" => PG::TextDecoder::TimestampWithTimeZone,
           }
 
-          known_coder_types = coders_by_name.keys.map { |n| quote(n) }
+          known_coder_types = coders_by_name.collect { |k, _v| quote(k) }
```

Comment:
Ditto here, it's not a hot path and the new version is harder to read because of the ignored block parameter.

If we really wanted to optimise this method, we could compute `coders_by_name` and `known_coder_types` once instead of each time it's called, since they're static data.

Reply:
Sorry, I didn't mean that I wanted you to apply the larger refactoring. My point was that making a tiny optimisation that saves a single allocation, when there's a (relatively) much bigger opportunity in the surrounding code, suggests that this is a micro-optimisation that won't have any noticeable effect.

I was fine with the `map` -> `map!` changes, since they're no less readable (in my opinion) and a few of them look like they could actually be called multiple times per request. This one was over the threshold where the new code looks worse _and_ it's clearly not a hot path, so I objected to it. But really, anything that isn't obviously part of a commonly-used API probably shouldn't be touched, unless a benchmark shows that it's necessary.

Reply:
thanks @eugeneius, have reverted changes from both areas of concern.

Reply:
@eugeneius ran some benchmarks for the changes:
```ruby
NON_EMPTY = [ :some_key ]
ATTRIBUTES = { id: 123, account_id: 1, title: "test", description: "test" }

Benchmark.memory do |x|
  x.report('map +') { NON_EMPTY + ATTRIBUTES.keys.map(&:to_s) }
  x.report('map! +') { NON_EMPTY + ATTRIBUTES.keys.map!(&:to_s) }
  x.compare!
end
```
```shell
Comparison:
              map! +:        312 allocated
               map +:        384 allocated - 1.23x more
```
```ruby
Benchmark.ips do |x|
  x.report('map +') { NON_EMPTY + ATTRIBUTES.keys.map(&:to_s) }
  x.report('map! +') { NON_EMPTY + ATTRIBUTES.keys.map!(&:to_s) }
  x.compare!
end
```
```shell
Warming up --------------------------------------
               map +   112.040k i/100ms
              map! +   110.965k i/100ms
Calculating -------------------------------------
               map +      1.086M (± 2.9%) i/s -      5.490M in   5.058600s
              map! +      1.138M (± 3.3%) i/s -      5.770M in   5.077784s

Comparison:
              map! +:  1137568.4 i/s
               map +:  1086188.1 i/s - same-ish: difference falls within error
```
```ruby
Benchmark.memory do |x|
  x.report('keys.include') { arr = ATTRIBUTES.keys.map(&:to_s); arr.include?(:idd) || arr.include?(:id) }
  x.report('.key?') { ATTRIBUTES.key?(:idd) || ATTRIBUTES.key?(:id) }
  x.compare!
end
```
```shell
Comparison:
               .key?:          0 allocated
        keys.include:        304 allocated - Infx more
```
```ruby
Benchmark.ips do |x|
  x.report('keys.include') { arr = ATTRIBUTES.keys.map(&:to_s); arr.include?(:idd) || arr.include?(:id) }
  x.report('.key?') { ATTRIBUTES.key?(:idd) || ATTRIBUTES.key?(:id) }
  x.compare!
end
```
```shell
Warming up --------------------------------------
        keys.include    58.724k i/100ms
               .key?   845.082k i/100ms
Calculating -------------------------------------
        keys.include    587.523k (± 3.1%) i/s -      2.936M in   5.002171s
               .key?      8.362M (± 2.1%) i/s -     42.254M in   5.055239s

Comparison:
               .key?:  8362499.2 i/s
        keys.include:   587523.1 i/s - 14.23x  (± 0.00) slower
```
`activerecord/lib/active_record/associations/association.rb#initialize_attributes` is called for every AR object that is built/created through some kind of an AR association.